File: Top_100_Liked_Questions/3._Longest_Substring_Without_Repeating_Characters/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # sliding window method O(n)
        # have pointers l and r
        # and a hash map for the latest index for the letter
        ans = l = 0
        hashMap = {}
        
        for r in range(len(s)):
            if s[r] in hashMap:
                l = max(hashMap[s[r]] + 1, l)
            ans = max(ans, r - l + 1)
            hashMap[s[r]] = r
        
        return ans

File: Top_100_Liked_Questions/3._Longest_Substring_Without_Repeating_Characters/solution.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let l = 0;
    let ans = 0;
    let latestIdx = {}
    
    for (let r=0; r<s.length; r++) {
        if (latestIdx.hasOwnProperty(s[r])) {
            l = Math.max(l, latestIdx[s[r]] + 1);
        }
        ans = Math.max(ans, r - l + 1);
        latestIdx[s[r]] = r;
    }
    
    return ans;
};